### PR TITLE
feat: Add ECLArchiveViewer

### PR DESCRIPTION
**Pull request by jbrundage**
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [ ] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

**Comment by jbrundage** [comment]
73AD54B0-D3CF-4A30-8AD4-3EBD6C59A03E

**Comment by jbrundage** [comment]
It works well with SplitPanel resizing:
A6868EBE-52A6-48AE-9B43-5D50824448DE

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+                this._rawArchiveCache = json;
+                this._archiveCache = this.transformArchiveData(json);
+                callback(json);
+            })
+            ;
+    }
+
+    viewRawArchive() {
+        console.log(this._rawArchiveCache);
+    }
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: remove or rename (logRawArchive).

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+            .then(r => r.text())
+            .then(xml => {
+                this._cacheUrl = url;
+                const json = xml2json(xml);
+                this._rawArchiveCache = json;
+                this._archiveCache = this.transformArchiveData(json);
+                callback(json);
+            })
+            ;
+    }
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: The fetch archive logic should be in @hpcc-js/comms

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+                + "-DL_Logs-DL_DetailAx0"
+                ;
+        }
+        if (this._archiveCache && url === this._cacheUrl) {
+            callback(this._archiveCache);
+        } else {
+            this.fetchArchives(url, callback);
+        }
+    }
+    fetchArchives(url, callback) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: No callback, use promise.

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            const archiveData = this.transformArchiveData(json);
+            console.log("archiveData", archiveData);
+            this._directoryPane
+                .data(archiveData)
+                .render()
+                ;
+        });
+    }
+
+    getArchives(callback) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: no callback, should use promise

**Comment by jbrundage** [comment]
@GordonSmith this has been revised

File: packages/common/package.json [hunk 5]
```diff
@@ -39,6 +39,7 @@
     "@types/d3-dsv": "1.0.36",
     "@types/d3-ease": "1.0.8",
     "@types/d3-format": "1.3.1",
+    "@types/d3-hierarchy": "1.1.6",
```
Comment (GordonSmith) [inline packages/common/package.json]: Should not be in common - the only d3 items in common are ones used by common...

File: packages/common/package.json [hunk 6]
```diff
@@ -60,6 +61,7 @@
     "d3-ease": "1.0.5",
     "d3-format": "1.3.2",
     "d3-interpolate": "1.3.2",
+    "d3-hierarchy": "1.1.9",
```
Comment (GordonSmith) [inline packages/common/package.json]: Should not be in common - the only d3 items in common are ones used by common...

File: packages/common/src/index.ts [hunk 7]
```diff
@@ -44,6 +44,7 @@ export * from "d3-drag";
 export * from "d3-dsv";
 export * from "d3-ease";
 export * from "d3-format";
+export * from "d3-hierarchy";
```
Comment (GordonSmith) [inline packages/common/src/index.ts]: Should not be in common - the only d3 items in common are ones used by common...

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 8]
```diff
@@ -0,0 +1,343 @@
+import { Editor } from "@hpcc-js/codemirror";
+import { hierarchy as d3Hierarchy, HTMLWidget, Palette, select as d3Select } from "@hpcc-js/common";
+// import {  } from "@hpcc-js/comms"; // TODO: move fetch logic to comms
+import { SplitPanel } from "@hpcc-js/phosphor";
+import { xml2json } from "@hpcc-js/util";
+
+class ArchiveList extends HTMLWidget {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: Move "ArchiveList" to @hpcc-js/tree

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            ;
+
+        rowEnter
+            .merge(rowSelection)
+            .style("background-color", context.backgroundColor())
+            ;
+
+        rowSelection.exit().remove();
+    }
+    getRowElementDepth(elm) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: For consistency move to above enter/update/exit methods (in general any helper methods used by those functions should be above them)

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+        const depthClass = (elm.attr("class").split(" ") as any).find(n => n.indexOf("directory-row-depth-") === 0);
+        return Number(depthClass.replace("directory-row-depth-", ""));
+    }
+    flattenData(json) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: For consistency move to above enter/update/exit methods

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                depth: node.depth,
+                content: node.data.content,
+                isFolder: !!node.data.children
+            });
+            if (node.children) {
+                node.children.forEach(visitNode);
+            }
+        }
+    }
+    determineIconClass(d) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: For consistency move to above enter/update/exit methods

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+                            rowDiv.classed("folder-closed", false);
+                            rowDiv.classed("folder-open", true);
+                            iconDiv.attr("class", "row-icon " + context.folderIconOpen());
+                        } else {
+                            rowDiv.classed("folder-closed", true);
+                            rowDiv.classed("folder-open", false);
+                            iconDiv.attr("class", "row-icon " + context.folderIconClosed());
+                        }
+                        while (next !== null) {
+                            const nextDepth = Number(context.getRowElementDepth(d3Select(next)));
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: `Number(...)` not needed?

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                                next = null;
+                            }
+                        }
+                    });
+                } else {
+                    rowDiv.on("click", () => {
+                        element.selectAll(".row-label").style("font-weight", "normal");
+                        labelDiv.style("font-weight", "bold");
+                        const ext = d.label.split(".").pop().toLowerCase();
+                        context.rowClick(ext === "json" ? JSON.stringify(JSON.parse(d.content), null, 4) : d.content);
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: I don't think the ArchiveList should be responsible for formatting the content (in fact it shouldn't really known about the content at all)...

File: packages/eclwatch/src/ECLArchiveViewer.ts (hunk 11 above)
Comment (GordonSmith) [reply packages/eclwatch/src/ECLArchiveViewer.ts]: Also mark as protected

Comment (GordonSmith) [reply packages/eclwatch/src/ECLArchiveViewer.ts]: ...and rename to "iconClass"

File: packages/eclwatch/src/ECLArchiveViewer.ts (hunk 9 above)
Comment (GordonSmith) [reply packages/eclwatch/src/ECLArchiveViewer.ts]: ...and mark as protected
...and rename to "elementDepth"

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+            const _data = this.transformArchiveTreeData(json);
+            this._directoryPane
+                .data(_data)
+                .render()
+                ;
+        }
+        this.relativeSizes([this.directoryWidthRatio(), 1 - this.directoryWidthRatio()]);
+    }
+
+    transformArchiveTreeData(json) {
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: For consistency move to above enter/update/exit methods
...and mark as protected

File: tests/test-eclwatch/src/eclwatch.spec.ts [hunk 15]
```diff
@@ -19,6 +19,8 @@ describe("@hpcc-js/eclwatch", function () {
                 }
                 if (item.prototype instanceof HTMLWidget || item.prototype instanceof SVGWidget) {
                     switch (item.prototype.constructor) {
+                        case ECLArchiveViewer:
+                            break;
```
Comment (GordonSmith) [inline tests/test-eclwatch/src/eclwatch.spec.ts]: Add actual tests

File: packages/common/package.json (hunk 5 above)
Comment (jbrundage) [reply packages/common/package.json]: I thought we were going to stop importing d3 modules in any packages other than common? I think you said it had something to do with typing? Maybe I misunderstood?

Comment (GordonSmith) [reply packages/common/package.json]: I don't believe I said that - there was an issue with d3-event having more than one instance?

File: packages/eclwatch/src/ECLArchiveViewer.ts (hunk 13 above)
Comment (jbrundage) [reply packages/eclwatch/src/ECLArchiveViewer.ts]: This is just a convenience feature that allows JSON to be formatted for easy readability. Maybe `codemirror_JSONEditor` should do this by default?

File: packages/tree/src/DirectoryTree.ts [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+                            ;
+                        labelDiv
+                            .style("background-color", null)
+                            .style("color", context.fontColor())
+                            ;
+                    })
+                    ;
+                if (d.isFolder) {
+                    rowDiv.on("click", function(ev, i, node) {
+                        const folderDepth = context.elementDepth(rowDiv);
```
Comment (GordonSmith) [inline packages/tree/src/DirectoryTree.ts]: change `rowDiv =>d3Select(this)`

File: packages/eclwatch/src/ECLArchiveViewer.ts [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+                .fetchArchive()
+                .then(resp => {
+
+                    this._contentStr = resp;
+                    this.updateDirectoryPane(this._contentStr);
+
+                    this._prevHash = hashSum({
+                        baseUrl: this.baseUrl(),
+                        wuid: this.wuid()
+                    });
```
Comment (GordonSmith) [inline packages/eclwatch/src/ECLArchiveViewer.ts]: Should be:
```
this._prevHash = hash;
```

Comment (GordonSmith) [reply packages/eclwatch/src/ECLArchiveViewer.ts]: (Especially if Wuid or baseUrl changed while waiting on the response...)

File: packages/tree/src/DirectoryTree.ts [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+export class DirectoryTree extends HTMLWidget {
+    constructor() {
+        super();
+    }
+
+    protected elementDepth(elm) {
+        const depthClass = (elm.attr("class").split(" ") as any).find(n => n.indexOf("directory-row-depth-") === 0);
+        return Number(depthClass.replace("directory-row-depth-", ""));
+    }
```
Comment (GordonSmith) [inline packages/tree/src/DirectoryTree.ts]: Removed unused code

**Comment by jbrundage** [comment]
@GordonSmith these items have been resolved

**Pull request merged**